File: core/src/ru/myitschool/dcrawler/event/EntityEventListener.java
package ru.myitschool.dcrawler.event;

import ru.myitschool.dcrawler.encounters.Encounter;
import ru.myitschool.dcrawler.entities.Entity;
import ru.myitschool.dcrawler.math.MathAction;

import java.util.HashMap;

public class EntityEventListener extends EventListener {

    public EntityEventListener() throws Exception {
        super();
    }

    @Override
    public HashMap<String, Object> fire(String eventCode, HashMap<String, Object> args) {
        HashMap<String, Object> result = new HashMap<>();
        Entity entity = (Entity) args.get(EntityEvent.ENTITY_EXECUTOR_ARG_KEY);
        switch (eventCode){
            case EntityEvent.START_TURN_EVENT:
                entity.startTurn();
                break;
            case EntityEvent.END_TURN_EVENT:
                entity.endTurn();
                break;
            case EntityEvent.START_MOVE_EVENT:
                entity.startMove();
                break;
            case EntityEvent.END_MOVE_EVENT:
                entity.endMove();
                break;
            case EntityEvent.COUNT_MP_EVENT:
                boolean withMovement = (boolean) args.get(EntityEvent.WITH_MOVEMENT_ARG_KEY);
                int mp = entity.countMp(withMovement);
                result.put(EntityEvent.MP_ARG_KEY, mp);
                break;
            case EntityEvent.CAN_USE_SKILL_EVENT:
                boolean canUseSkill = entity.canUseSkill();
                result.put(EntityEvent.CAN_USE_SKILL_ARG_KEY, canUseSkill);
                break;
            case EntityEvent.START_SKILL_EVENT:
                entity.startSkill();
                break;
            case EntityEvent.END_SKILL_EVENT:
                entity.endSkill();
                break;
            case EntityEvent.ATTACK_BONUS_EVENT:
                MathAction action = (MathAction) args.get(EntityEvent.ATTACK_BONUS_ARG_KEY);
                action = entity.attackBonus(action);
                result.put(EntityEvent.ATTACK_BONUS_ARG_KEY, action);
                break;
            case EntityEvent.ACCURACY_BONUS_EVENT:
                int accuracy = (int) args.get(EntityEvent.ACCURACY_BONUS_ARG_KEY);
                Entity target = (Entity) args.get(EntityEvent.ENTITY_TARGET_ARG_KEY);
                accuracy = entity.accuracyBonus(accuracy, target);
                result.put(EntityEvent.ACCURACY_BONUS_ARG_KEY, accuracy);
                break;
            case EntityEvent.ON_DAMAGE_EVENT:
                int damage = (int) args.get(EntityEvent.DAMAGE_ARG_KEY);
                damage = entity.onDamage(damage);
                result.put(EntityEvent.DAMAGE_ARG_KEY, damage);
                break;
            case EntityEvent.ON_HEAL_EVENT:
                int heal = (int) args.get(Entity.HEAL_ARG_KEY);
                heal = entity.onHeal(heal);
                result.put(EntityEvent.HEAL_ARG_KEY, heal);
                break;
            case EntityEvent.ON_ENCOUNTER_EVENT:
                Encounter encounter = (Encounter) args.get(EntityEvent.ENCOUNTER_ARG_KEY);
                entity.onEncounter(encounter);
                break;
            case EntityEvent.ON_DEATH_EVENT:
                entity.onDeath();
                break;
        }
        return result;
    }

    @Override
    public String[] eventCodes() {
        String[] codes = {EntityEvent.START_TURN_EVENT, EntityEvent.END_TURN_EVENT, EntityEvent.START_MOVE_EVENT,
                EntityEvent.END_MOVE_EVENT, EntityEvent.COUNT_MP_EVENT, EntityEvent.CAN_USE_SKILL_EVENT,
                EntityEvent.START_SKILL_EVENT, EntityEvent.END_SKILL_EVENT, EntityEvent.ATTACK_BONUS_EVENT,
                EntityEvent.ACCURACY_BONUS_EVENT, EntityEvent.ON_DAMAGE_EVENT, EntityEvent.ON_HEAL_EVENT,
                EntityEvent.ON_ENCOUNTER_EVENT, EntityEvent.ON_DEATH_EVENT};
        return codes;
    }
}

File: core/src/com/wolg_vlad/dcrawler/event/EntityEventListener.kt
package com.wolg_vlad.dcrawler.event

import com.wolg_vlad.dcrawler.encounters.Encounter
import com.wolg_vlad.dcrawler.entities.Entity
import com.wolg_vlad.dcrawler.math.MathAction
import java.util.*

class EntityEventListener : EventListener() {
    override fun fire(eventCode: String?, args: HashMap<String, Any>): HashMap<String, Any> {
        val result = HashMap<String, Any>()
        val entity = args[EntityEvent.ENTITY_EXECUTOR_ARG_KEY] as Entity?
        when (eventCode) {
            EntityEvent.START_TURN_EVENT -> entity!!.startTurn()
            EntityEvent.END_TURN_EVENT -> entity!!.endTurn()
            EntityEvent.START_MOVE_EVENT -> entity!!.startMove()
            EntityEvent.END_MOVE_EVENT -> entity!!.endMove()
            EntityEvent.COUNT_MP_EVENT -> {
                val withMovement = args[EntityEvent.WITH_MOVEMENT_ARG_KEY] as Boolean
                val mp = entity!!.countMp(withMovement)
                result[EntityEvent.MP_ARG_KEY] = mp
            }
            EntityEvent.CAN_USE_SKILL_EVENT -> {
                val canUseSkill = entity!!.canUseSkill()
                result[EntityEvent.CAN_USE_SKILL_ARG_KEY] = canUseSkill
            }
            EntityEvent.START_SKILL_EVENT -> entity!!.startSkill()
            EntityEvent.END_SKILL_EVENT -> entity!!.endSkill()
            EntityEvent.ATTACK_BONUS_EVENT -> {
                var attackAction = args[EntityEvent.ATTACK_BONUS_ARG_KEY] as MathAction
                attackAction = entity!!.attackBonus(attackAction)
                result[EntityEvent.ATTACK_BONUS_ARG_KEY] = attackAction
            }
            EntityEvent.HEAL_BONUS_ARG_KEY -> {
                var healAction = args[EntityEvent.HEAL_BONUS_ARG_KEY] as MathAction?
                healAction = entity!!.healBonus(healAction!!)
                result[EntityEvent.HEAL_BONUS_ARG_KEY] = healAction
            }
            EntityEvent.ACCURACY_BONUS_EVENT -> {
                var accuracy = args[EntityEvent.ACCURACY_BONUS_ARG_KEY] as Int
                val target = args[EntityEvent.ENTITY_TARGET_ARG_KEY] as Entity?
                accuracy = entity!!.accuracyBonus(accuracy, target!!)
                result[EntityEvent.ACCURACY_BONUS_ARG_KEY] = accuracy
            }
            EntityEvent.ON_DAMAGE_EVENT -> {
                var damage = args[EntityEvent.DAMAGE_ARG_KEY] as Int
                damage = entity!!.onDamage(damage)
                result[EntityEvent.DAMAGE_ARG_KEY] = damage
            }
            EntityEvent.ON_HEAL_EVENT -> {
                var heal = args[EntityEvent.HEAL_ARG_KEY] as Int
                heal = entity!!.onHeal(heal)
                result[EntityEvent.HEAL_ARG_KEY] = heal
            }
            EntityEvent.ON_ENCOUNTER_EVENT -> {
                val encounter = args[EntityEvent.ENCOUNTER_ARG_KEY] as Encounter
                entity!!.onEncounter(encounter)
            }
            EntityEvent.ON_DEATH_EVENT -> entity!!.onDeath()
        }
        return result
    }

    override fun eventCodes(): Array<String> {
        return arrayOf(EntityEvent.START_TURN_EVENT, EntityEvent.END_TURN_EVENT, EntityEvent.START_MOVE_EVENT,
                EntityEvent.END_MOVE_EVENT, EntityEvent.COUNT_MP_EVENT, EntityEvent.CAN_USE_SKILL_EVENT,
                EntityEvent.START_SKILL_EVENT, EntityEvent.END_SKILL_EVENT, EntityEvent.ATTACK_BONUS_EVENT,
                EntityEvent.ACCURACY_BONUS_EVENT, EntityEvent.ON_DAMAGE_EVENT, EntityEvent.ON_HEAL_EVENT,
                EntityEvent.ON_ENCOUNTER_EVENT, EntityEvent.ON_DEATH_EVENT)
    }
}